### PR TITLE
Refactor KNN graph handling and outlier detection in issue managers

### PR DESCRIPTION
**Pull request by elisno**
## Summary

🎯 **Purpose**: This PR refactors the handling of KNN graphs in various issue managers, including a significant update to the outlier issue manager. The changes aim to improve flexibility, consistency, and error handling across Datalab's issue detection mechanisms.

## Key Changes

1. Introduced helper functions for KNN graph processing in knn_graph_helpers.py.

2. Replaced existing methods with these helpers in multiple issue managers.

3. Added checks to validate the number of neighbors in stored KNN graphs,
user-provided ones should not require such validation.

4. Significantly refactored the OutlierIssueManager:

   - Simplified API by replacing NearestNeighbors object with direct parameters.

   - Restructured find_issues method for better handling of different input scenarios.

   - Improved KNN graph management and outlier score computation.

   - Updated related test cases to reflect new (intended) behavior.

## Impact

🌐 Areas Affected:

- ```
  cleanlab/datalab/internal/issue_manager/
  ├── data_valuation.py
  ├── duplicate.py
  ├── noniid.py
  ├── outlier.py
  └── underperforming_group.py
  ```
- Added knn_graph_helpers.py for shared KNN graph functionality.


👥 Who’s Affected:


- Users who expect Datalab to automatically lower the number of neighbors `k` will now receive an error.

- The OutlierIssueManager now has a more consistent API with other issue managers.

- Users providing pre-computed KNN graphs are now responsible for tracking their own metrics and values of k.

## Testing

🔍 Testing Done: 
- Updated existing tests to reflect changes in KNN graph handling and outlier detection.
- Added new tests to ensure proper validation and error handling for insufficient KNN graphs.
- Verified that all tests in `tests/datalab/` pass locally.

## Future Considerations

- The changes to the OutlierIssueManager lay the groundwork for potential future optimizations, such as storing NearestNeighbors objects in info["statistics"] for global reuse across issue managers.
- Any edge cases discovered in further use will be addressed in future updates.

## Notes

- These changes represent a significant step towards more consistent and flexible issue detection in Datalab, particularly for outlier detection.
- The PR aims to balance immediate improvements with the potential for future enhancements, prioritizing incremental improvements to the library.

### CODE REVIEW

**Comment by jwmueller** [comment]
can you ask @sanjanag for review? Thanks!

File: cleanlab/datalab/internal/issue_manager/knn_graph_helpers.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+from cleanlab.typing import Metric
+
+
+def num_neighbors_in_knn_graph(knn_graph: csr_matrix) -> int:
+    """Calculate the number of neighbors per row in a knn graph."""
+    return knn_graph.nnz // knn_graph.shape[0]
+
+
+def _process_knn_graph_from_inputs(
+    kwargs: Dict[str, Any], statistics: Dict[str, Any], k_for_recomputation: int
```
Comment (sanjanag) [inline cleanlab/datalab/internal/issue_manager/knn_graph_helpers.py:17]: Rename kwargs

File: cleanlab/datalab/internal/issue_manager/noniid.py [hunk 2]
```diff
@@ -145,42 +146,38 @@ def _determine_features(
...
     ) -> None:
-        knn_graph = self._process_knn_graph_from_inputs(kwargs)
-        old_knn_metric = self.datalab.get_info("statistics").get("knn_metric")
-        metric_changes = bool(self.metric and self.metric != old_knn_metric)
+        statistics = self.datalab.get_info("statistics")
 
-        if knn_graph is None or metric_changes:
-            if features is None and pred_probs is not None:
-                self._skip_storing_knn_graph_for_pred_probs = True
+        self._check_pred_probs(features, pred_probs, kwargs, statistics, self.k)
```
Comment (sanjanag) [inline cleanlab/datalab/internal/issue_manager/noniid.py:172]: Add a comment for this condition

File: cleanlab/datalab/internal/issue_manager/noniid.py [hunk 3]
```diff
@@ -208,27 +205,18 @@ def find_issues(
 
         self.info = self.collect_info(knn_graph=knn_graph)
 
-    def _process_knn_graph_from_inputs(self, kwargs: Dict[str, Any]) -> Union[csr_matrix, None]:
-        """Determine if a knn_graph is provided in the kwargs or if one is already stored in the associated Datalab instance."""
-        knn_graph_kwargs: Optional[csr_matrix] = kwargs.get("knn_graph", None)
-        knn_graph_stats = self.datalab.get_info("statistics").get("weighted_knn_graph", None)
+    def _check_pred_probs(self, features, pred_probs, kwargs, statistics, k) -> None:
```
Comment (sanjanag) [inline cleanlab/datalab/internal/issue_manager/noniid.py:208]: Rename method

File: cleanlab/datalab/internal/issue_manager/outlier.py [hunk 4]
```diff
@@ -15,21 +15,22 @@
 # along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any, ClassVar, Dict, Optional, Tuple, Union, cast
+from typing import TYPE_CHECKING, Any, ClassVar, Dict, Optional, Tuple
```
Comment (sanjanag) [inline cleanlab/datalab/internal/issue_manager/outlier.py:18]: Remove score_with_features

File: cleanlab/datalab/internal/issue_manager/outlier.py [hunk 5]
```diff
@@ -156,6 +167,15 @@ def find_issues(
             )  # Useful info for detecting issues in test data
             is_issue_column = scores < issue_threshold
 
+        else:
+            # Handle case where neither kNN nor pred_probs-based detection is possible
+            if kwargs.get("knn_graph", None) is not None:
```
Comment (sanjanag) [inline cleanlab/datalab/internal/issue_manager/outlier.py:172]: This should also check knn_graph in statistics

**Review by sanjanag — APPROVED** [review]

**Comment by sanjanag** [comment]
LGTM, merge whenever the comments are in.

File: cleanlab/datalab/internal/issue_manager/knn_graph_helpers.py (hunk 1 above)
Comment (elisno) [reply cleanlab/datalab/internal/issue_manager/knn_graph_helpers.py]: Renamed it to `user_find_issues_kwargs` for extra clarity. Renamed some of the variables accordingly.

**Pull request merged**